### PR TITLE
TaskKeysTableView

### PR DESCRIPTION
**Pull request by shaobinspire**
I have put a demo to demos/task_key_table_view.
The functions `populate_xxx` in KeyBindingsWindow.cpp are used for testing.

I am currently facing the challenge in adding a PopupBox to the EditableTableView. In the current implementation, the RegionBox is wrapped by the AnyInputBox, which in turn is wrapped by the EditableBox. This EditableBox is then passed to the ViewBuilder of the EditableTableView to construct a cell. However, this structure presents difficulties in integrating a PopupBox seamlessly.

If the PopupBox encapsulates the RegionBox directly, a compatibility issue arises: the AnyInputBox cannot accept the PopupBox since the PopupBox lacks the necessary interfaces such as the Submit signal and read-only related interfaces. Although one potential solution to implement a PopupRegionBox can address this interface compatible issue, it poses another challenge. When EditableBox wraps the PopupRegionBox, the act of popping up the PopupRegionBox will trigger a editing end signal, as the popped component gets focus, so the RegionBox can’t trigger the editing state in the popup status.

Alternatively, if the PopupBox wraps the EditableBox, it could address the issues mentioned above. However this approach will break the signature of the ViewBuilder. The ViewBuilder in EditableTableView returns the pointer to the EditableBox, not a pointer to the QWidget.

### CODE REVIEW

**Review by kranar — CHANGES_REQUESTED** [review]
Still needs a lot of work.

File: Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp [hunk 1]
```diff
@@ -1,21 +1,77 @@
...
 #include "Spire/Ui/ScrollBox.hpp"
+#include "Spire/Ui/TableView.hpp"
 
 using namespace boost::signals2;
 using namespace Nexus;
 using namespace Spire;
 using namespace Spire::Styles;
 
+namespace {
+  auto populate_order_task_arguments() {
```
[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp:25]: This function is used for testing and will be removed after testing.

File: Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp [hunk 2]
```diff
@@ -1,21 +1,77 @@
...
+    arguments->push({"Test3",
+      Region(DefaultCountries::US()), "NYSE",
+      OrderType::MARKET, Side::BID, Quantity(20),
+      TimeInForce(TimeInForce::Type::IOC), {}, QKeySequence("F3")});
+    return arguments;
+  }
+
+  auto populate_region_box_model() {
+    auto securities = std::vector<std::pair<std::string, std::string>>{
+      {"MSFT.NSDQ", "Microsoft Corporation"},
```
[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp:44]: This function is same as the function `populate_order_task_arguments` and is used for testing.

File: Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp:25]: Never mix tests with implementation.

[the remark this answers is not in the document]
Comment (shaobinspire) [reply Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp:25]: Where can I populate testing data? In the KeyBindingsModel?

[the remark this answers is not in the document]
Comment (kranar) [reply Applications/Spire/Source/KeyBindings/KeyBindingsWindow.cpp:25]: Test code and implementations are separate. Identify which test code is most appropriate to populate your model there. I don't know off the top of my head where that is, I can look later tonight when I get some time.

File: Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+  }
+
+  connection connect_operation_signal(
+      const OperationSignal::slot_type& slot) const override {
+    return m_transaction.connect_operation_signal(slot);
+  }
+
+  void on_operation(const OrderTaskArgumentsListModel::Operation& operation) {
+    visit(operation,
+      [&] (const OrderTaskArgumentsListModel::AddOperation& operation) {
```
Comment (kranar) [inline Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp:319]: To properly handle transactions you'll need to add the following cases:

    [&] (const StartTransaction&) {
      m_transaction.start();
    },
    [&] (const EndTransaction&) {
      m_transaction.end();
    },

File: Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+      [&] (const OrderTaskArgumentsListModel::MoveOperation& operation) {
+        m_transaction.push(TableModel::MoveOperation(
+          operation.m_source, operation.m_destination));
+      },
+      [&] (const OrderTaskArgumentsListModel::RemoveOperation& operation) {
+        m_transaction.push(TableModel::RemoveOperation(operation.m_index,
+          to_list(operation.get_value())));
+      },
+      [&] (const OrderTaskArgumentsListModel::UpdateOperation& operation) {
+        for(auto i = 0; i < COLUMN_SIZE; ++i) {
```
Comment (kranar) [inline Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp:332]: To start a transaction you need to call ```m_transaction.transact(...)```.

File: Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+      return new AnyInputBox(*make_label(""));
+    } else {
+      return new AnyInputBox(*new KeyInputBox(
+        make_validated_value_model<QKeySequence>(&key_input_box_validator,
+          to_value_model<QKeySequence>(table, row, column))));
+    } 
+  }();
+  if(column_id == Column::KEY) {
+    return new EditableBox(*input_box, [] (const auto& key) {
+      return key_input_box_validator(key) == QValidator::Acceptable;
```
Comment (kranar) [inline Applications/Spire/Source/KeyBindings/TaskKeysTableView.cpp:388]: It's more robust to test:
    ```test != QValidator::Invalid```

**Review by kranar — CHANGES_REQUESTED** [review]

**Comment by kranar** [comment]
Is it possible to subclass ```PopupBox``` into a ```PopupInputBox``` that adds the methods needed like ```get_current```, ```get_submission```, etc... so that it can plug into an ```AnyInputBox```?

**Comment by shaobinspire** [comment]
> PopupInputBox

I think it might work. If the EditableBox encapsulates the PopupBox, the popup of the PopupBox will lead to the EditableBox losing focus and reverting to the non-editable state. Unless the PopupInputBox prevents it from being set to read-only on popup. However, if the PopupBox wraps the EditableBox instead, the EditableBox becomes the popup component and it would retain focus for editing state. Hence, I think it's more intuitive for PopupBox to wrap EditableBox.

**Comment by kranar** [comment]
Perhaps you can subclass the EditableBox into an EditablePopupBox.

**Comment by shaobinspire** [comment]
> Perhaps you can subclass the EditableBox into an EditablePopupBox.

Good point. I will give it a try.

**Comment by kranar** [comment]
Was this reviewed by Jon?

**Comment by shaobinspire** [comment]
> Was this reviewed by Jon?

No. I am going to assign it to Jon for review.

**Comment by shaobinspire** [comment]
@kranar I have fixed [the task](https://app.asana.com/0/1169210941620249/1207078420458093/f) and updated the code based on code review.

**Review by kranar — APPROVED** [review]
Looks good!

**Pull request merged**